File: docs/Blocks/BlockGame.py
import sys

N = int(input())
T = int(input())
C = int(input())
F = float(input())
P = float(input())

# read grid
grid = [[0 for x in range(N)] for y in range(N)]
for r in range(N):
  for c in range(N):
    grid[r][c] = int(input())

# read tiles
S=3;
tiles = [[[0 for x in range(S)] for y in range(S)] for z in range(T)]
for i in range(T):
  line=input()
  for r in range(S):
    for c in range(S):
      tiles[i][r][c]=ord(line[r*S+c])-ord('0')
  

q=N//S

for i in range(1000):        
  Id=i%T;
  r=((i//q)*S)%(N-2);
  c=((i%q)*S)%(N-2);
  
  # check for a valid move
  valid = True
  for r2 in range(S):
    for c2 in range(S):
      if tiles[Id][r2][c2]>0 and grid[r+r2][c+c2]>0: valid = False
  if not(valid): continue

  # place tile
  for r2 in range(S):
    for c2 in range(S):  
      if tiles[Id][r2][c2]>0:
        grid[r+r2][c+c2] = tiles[Id][r2][c2]
        
  # print move
  print(str(Id)+" "+str(r)+" "+str(c))
  sys.stdout.flush()
  
  # read tile
  line=input()
  for r2 in range(S):
    for c2 in range(S):    
      tiles[Id][r2][c2]=ord(line[r2*S+c2])-ord('0')
    
  elapsedTime = int(input())

# terminate
print("-1");
sys.stdout.flush()

File: docs/Blocks/BlockGame.cpp
#include <vector>
#include <iostream>
using namespace std;

int main() 
{
  int N, T, C;
  double F, P;
  cin >> N >> T >> C >> F >> P;

  //read grid
  int grid[N][N];
  for (int r=0; r<N; r++)
    for (int c=0; c<N; c++)
    {
      cin >> grid[r][c];
    }

  //read tiles
  const int S=3;
  int tiles[T][S][S];
  for (int i=0; i<T; i++)
  {
    string line;
    cin >> line;
    for (int r=0; r<S; r++)
      for (int c=0; c<S; c++)
        tiles[i][r][c] = (int)(line[r*S+c]-'0');
  }

  int q=N/S;
          
  for (int i=0; i<1000; i++)
  {
    int id=i%T;
    int r=((i/q)*S)%(N-2);
    int c=((i%q)*S)%(N-2);
    
    // check for a valid move
    bool valid = true;
    for (int r2=0; r2<S; r2++)
      for (int c2=0; c2<S; c2++)
        if (tiles[id][r2][c2]>0 && grid[r+r2][c+c2]>0) valid = false;
    if (!valid) continue;

    // place tile
    for (int r2=0; r2<S; r2++)
      for (int c2=0; c2<S; c2++)
        if (tiles[id][r2][c2]>0)
          grid[r+r2][c+c2] = tiles[id][r2][c2];

    //print move
    cout << id << " " << r << " " << c << endl;
    cout.flush();
    
    //read tile
    string line;
    cin >> line;
    for (int r2=0; r2<S; r2++)
      for (int c2=0; c2<S; c2++)
        tiles[id][r2][c2] = (int)(line[r2*S+c2]-'0');
      
    int elapsedTime;
    cin >> elapsedTime;
  }

  //terminate
  cout << "-1" << endl;
  cout.flush();
}
